### PR TITLE
Modify S4423[Python]: Support of AWS CDK code (APPSEC-103)

### PR DESCRIPTION
**Pull request by pierre-loup-tristant-sonarsource**
Specification ticket: [APPSEC-103](https://sonarsource.atlassian.net/browse/APPSEC-103)
Implementation ticket: [SONARPY-1127](https://sonarsource.atlassian.net/browse/SONARPY-1127)
Preview: [S4423](https://sonarsource.github.io/rspec/#/rspec/S4423/python)

### CODE REVIEW

File: rules/S4423/python/message.adoc [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+Change this code to use a stronger protocol.
+
+==== AWS APIGateway
+
+Change this code to enforce TLS 1.2 or above.
+
+==== AWS OpenSearch / Elasticsearch
+
+Omitting "tls_security_polic" enables deprecated version of TLS. Set it to enforce TLS 1.2 or above.
```
Comment (egon-okerman-sonarsource) [inline rules/S4423/python/message.adoc:13]:
```suggestion
Omitting "tls_security_policy" enables a deprecated version of TLS. Set it to enforce TLS 1.2 or above.
```

Small typo

File: rules/S4423/python/rule.adoc [hunk 2]
```diff
@@ -23,31 +16,85 @@ import ssl
...
+[source,python]
+----
+from aws_cdk.aws_apigateway import DomainName, SecurityPolicy
+class ExampleStack(Stack):
+    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
+        super().__init__(scope, construct_id, **kwargs) 
+        DomainName(self, "example",
+            domain_name="example.com",
+            certificate=certificate,
+            security_policy=apigateway.SecurityPolicy.TLS_1_0 # Noncompliant
```
Comment (egon-okerman-sonarsource) [inline rules/S4423/python/rule.adoc:29]:
```suggestion
            security_policy=SecurityPolicy.TLS_1_0 # Noncompliant
```

`SecurityPolicy` is imported directly, so no need to reference a module.

File: rules/S4423/python/rule.adoc [hunk 3]
```diff
@@ -23,31 +16,85 @@ import ssl
 ssl.SSLContext(ssl.PROTOCOL_SSLv3) # Noncompliant
 ----
 
+For  https://docs.aws.amazon.com/cdk/api/v1/python/aws_cdk.aws_apigateway/DomainName.html[aws_cdk.aws_apigateway.DomainName]:
```
Comment (egon-okerman-sonarsource) [inline rules/S4423/python/rule.adoc:19]:
```suggestion
For https://docs.aws.amazon.com/cdk/api/v1/python/aws_cdk.aws_apigateway/DomainName.html[aws_cdk.aws_apigateway.DomainName]:
```

Remove the extra space

File: rules/S4423/python/rule.adoc [hunk 4]
```diff
@@ -23,31 +16,85 @@ import ssl
...
+[source,python]
+----
+from aws_cdk.aws_apigateway import DomainName, SecurityPolicy
+class ExampleStack(Stack):
+    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
+        super().__init__(scope, construct_id, **kwargs) 
+        DomainName(self, "example",
+            domain_name="example.com",
+            certificate=certificate,
+            security_policy=apigateway.SecurityPolicy.TLS_1_2
```
Comment (egon-okerman-sonarsource) [inline rules/S4423/python/rule.adoc:71]:
```suggestion
            security_policy=SecurityPolicy.TLS_1_2
```

`SecurityPolicy` is imported directly, so no need to reference a module.

File: rules/S4423/python/rule.adoc [hunk 5]
```diff
@@ -23,31 +16,85 @@ import ssl
...
+            domain_name="example.com",
+            certificate=certificate,
+            security_policy=apigateway.SecurityPolicy.TLS_1_0 # Noncompliant
+        )
+----
+
+For https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_opensearchservice/CfnDomain.html[aws_cdk.aws_opensearchservice.CfnDomain]:
+[source,python]
+----
+from aws_cdk.aws_opensearchservice import Domain, EngineVersion
```
Comment (egon-okerman-sonarsource) [inline rules/S4423/python/rule.adoc:36]:
```suggestion
from aws_cdk.aws_opensearchservice import CfnDomain, EngineVersion
```

Should be `CfnDomain`.

File: rules/S4423/python/rule.adoc [hunk 6]
```diff
@@ -23,31 +16,85 @@ import ssl
...
+class ExampleStack(Stack):
+    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
+        super().__init__(scope, construct_id, **kwargs)
+        Domain(self, "example",
```
Comment (egon-okerman-sonarsource) [inline rules/S4423/python/rule.adoc:40]:
```suggestion
        CfnDomain(self, "example",
```

Should be `CfnDomain`.

File: rules/S4423/python/rule.adoc [hunk 7]
```diff
@@ -23,31 +16,85 @@ import ssl
...
```
Comment (egon-okerman-sonarsource) [inline rules/S4423/python/rule.adoc:82]:
```suggestion
        CfnDomain(self, "example",
```

Should be `CfnDomain`

File: rules/S4423/python/rule.adoc [hunk 8]
```diff
@@ -23,31 +16,85 @@ import ssl
...
+            domain_name="example.com",
+            certificate=certificate,
+            security_policy=apigateway.SecurityPolicy.TLS_1_2
+        )
+----
+
+For https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_opensearchservice/CfnDomain.html[aws_cdk.aws_opensearchservice.CfnDomain]:
+[source,python]
+----
+from aws_cdk.aws_opensearchservice import Domain, EngineVersion
```
Comment (egon-okerman-sonarsource) [inline rules/S4423/python/rule.adoc:78]:
```suggestion
from aws_cdk.aws_opensearchservice import CfnDomain, EngineVersion
```

Should be `CfnDomain`

File: rules/S4423/python/rule.adoc [hunk 9]
```diff
@@ -23,31 +16,85 @@ import ssl
...
+class ExampleStack(Stack):
+    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
+        super().__init__(scope, construct_id, **kwargs)
+        Domain(self, "example",
+            version=EngineVersion.OPENSEARCH_1_3
+            tls_security_policy=opensearch.TLSSecurityPolicy.TLS_1_2
```
Comment (egon-okerman-sonarsource) [inline rules/S4423/python/rule.adoc:84]:
```suggestion
            domain_endpoint_options=CfnDomain.DomainEndpointOptionsProperty(
                tls_security_policy="Policy-Min-TLS-1-2-2019-07" # Compliant
            )
```

The parameter `tls_security_policy` does not exist here, and should instead be defined as shown above.

**Review by egon-okerman-sonarsource — CHANGES_REQUESTED** [review]

**Review by egon-okerman-sonarsource — CHANGES_REQUESTED** [review]
A few changes still have to be applied, but otherwise LGTM

**Review by egon-okerman-sonarsource — APPROVED** [review]

**Comment by pierre-loup-tristant-sonarsource** [comment]
Merged by accident. Reverted change and opened https://github.com/SonarSource/rspec/pull/1278

**Pull request merged**